### PR TITLE
Add all undeclared packages to `package.json`

### PR DESCRIPTION
**Pull request by smorimoto**
### What does it do?

**TL;DR**: The goal of this PR is to add (as many as possible) dependencies that are not listed.

**Glossary**:

- Safe Package
  - Safe Package is a package that has only one pinned version in the workspace.
- Unsafe Package
  - An unsafe package is a package that has multiple versions in the workspace.

**References**:

- Safe Packages
  - https://gist.github.com/smorimoto/6cac28bcdce5976f0ba42e8a5be2c417
    - `@strapi/design-system@1.6.3`
    - `@strapi/icons@1.6.3`
    - `@testing-library/react-hooks@8.0.1`
    - `@testing-library/react@12.1.4`
    - `formik@2.2.9`
    - `history@^4.9.0`
    - `immer@9.0.19`
    - `msw@1.0.0`
    - `react-dnd@15.1.2`
    - `react-helmet@^6.1.0`
    - `react-intl@6.2.7`
    - `react-query@3.24.3`
    - `react-redux@8.0.5`
    - `react-router-dom@5.3.4`
    - `react-select@5.6.0`
    - `react@^17.0.2`
    - `sqlite3@5.1.2`
    - `styled-components@5.3.3`
- Unsafe Packages
  - https://gist.github.com/smorimoto/07f7df255735bb250d3116565261f24d
    - `axios@1.2.2 | ^1.0.0 | ^0.26.0`
    - `fs-extra@^10.1.0 | 10.1.0 | 10.0.1 | ^10.0.0 | 10.0.0 | ^9.1.0 | ^9.0.1 | ^9.0.0 | ^8.1.0`
    - `match-sorter@^6.0.2 | ^4.0.2`
    - `pluralize@^8.0.0 | 8.0.0`
    - `prop-types@^15.8.1 | ^15.7.2 | ^15.6.2 | ^15.6.0 | ^15.0.0`
    - `qs@^6.11.0 | 6.11.0 | ^6.10.3 | 6.10.3 | ^6.10.2 | ^6.10.0 | ^6.9.6 | ^6.5.2 | ~6.5.2 | ^6.4.0`
    - `redux@^4.2.1 | ^4.1.2 | ^4.1.1`
    - `yup@^0.32.9 | 0.32.9`

**Strategies**:
- Safe Package
  - Check if the package is used only for the development purpose in the scope, then add it as a development dependency if it's used only for the development purpose; otherwise, add it as a dependency.
- Unsafe Package
  - Adding unsafe packages is not always that simple, but it depends on the distance between versions. If the distance
 between versions is minor, the program can fix it automatically; otherwise, humans would have to review it.

### Why is it needed?

Node allows you to import any package without having a version specified in your package.json. It can lead to subtle and hard-to-solve bugs.

For example:

- Node might find a globally installed package, and the project works on your machine. While other coworkers might be missing the globally installed package or (worse) have an incompatible version installed.

- Or Node might find a transitive dependency (dependency of a dependency) and use that. If you remove or upgrade that dependency and it affects the transitive dependency, it can trigger all sorts of bugs.

### How to test it?

Testing in CI is enough because it just technically adds dependency. (Well, this idea strongly assumes that Strapi's tests have reasonable coverage.)

### Related issue(s)/PR(s)

- https://github.com/strapi/strapi/issues/15094

(Let me link more later...)

### CODE REVIEW

**Comment by smorimoto** [comment]
This PR is ready for review now.

**Comment by smorimoto** [comment]
My team has been using Strapi for a while and encountered some problems, so I wrote a fixer to solve this from scratch and processed them all automatically. So no errors should be made (If it's bug-free!), but let's see if CI complains...

**Comment by smorimoto** [comment]
Great! CI seems to have no complaints.

File: packages/core/helper-plugin/package.json [hunk 1]
```diff
@@ -46,8 +46,10 @@
     "axios": "1.2.2",
     "date-fns": "2.29.3",
     "formik": "^2.2.6",
+    "fs-extra": "10.0.0",
```
Comment (joshuaellis) [inline packages/core/helper-plugin/package.json:49]: this should be in the devDeps as its used in development to generate the build index (although we no longer need this so it can be removed – should be a separate PR though).

File: packages/generators/generators/package.json [hunk 2]
```diff
@@ -30,13 +30,18 @@
...
+    "axios": "1.2.2",
     "chalk": "4.1.2",
     "fs-extra": "10.0.0",
     "node-plop": "0.26.3",
     "plop": "2.7.6",
-    "pluralize": "8.0.0"
+    "pluralize": "8.0.0",
+    "prop-types": "^15.7.2",
+    "react": "^17.0.2",
+    "react-router-dom": "5.3.4"
```
Comment (joshuaellis) [inline packages/generators/generators/package.json:44]: These are an interesting one too. The `files` are just copied over when someone makes a new plugin, how come you think their deps need to be declared? 😄

File: packages/plugins/i18n/package.json [hunk 3]
```diff
@@ -24,12 +24,27 @@
...
+    "react-redux": "8.0.5",
+    "react-router-dom": "5.3.4",
+    "styled-components": "5.3.3",
+    "yup": "^0.32.9"
   },
   "devDependencies": {
     "@testing-library/react": "12.1.4",
-    "msw": "1.0.0"
+    "msw": "1.0.0",
+    "redux": "^4.2.1"
```
Comment (joshuaellis) [inline packages/plugins/i18n/package.json:47]: react-redux requires a peer dep of redux, so i would actually add this to the `dependencies` :)

**Review by joshuaellis — CHANGES_REQUESTED** [review]
Overall looks like great work. I think the `plugins` templates need to be updated so when people are generating them it adds the correct deps – I can do this in a follow up PR.

Some questions inline to be addressed and a final note – can we revert the `bin` changes please?

File: packages/core/database/package.json [hunk 4]
```diff
@@ -37,6 +37,7 @@
     "knex": "2.4.0",
     "lodash": "4.17.21",
     "semver": "7.3.8",
+    "sqlite3": "5.1.2",
```
[the remark this answers is not in the document]
Comment (smorimoto) [reply packages/core/database/package.json:40]: It seems to have been added because it's used here. I've automated and fixed all this, so it might be weird in some places.
https://github.com/strapi/strapi/blob/78630d6da3e397cc4f3ecf075bc777dcb643addb/packages/core/database/lib/connection.js#L20-L24

Log: https://gist.github.com/smorimoto/6cac28bcdce5976f0ba42e8a5be2c417#file-safe-packages-json-L19

[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/database/package.json:40]: Reading the code we can revert it to not being included. We test to see if it's installed before using that class and the class is also not exported 😊

Thanks for the explanation!

File: packages/core/helper-plugin/package.json (hunk 1 above)
Comment (smorimoto) [reply packages/core/helper-plugin/package.json:49]: This PR https://github.com/strapi/strapi/pull/15797 lets us remove this completely! Fantastic!

Related log: https://gist.github.com/smorimoto/07f7df255735bb250d3116565261f24d#file-unsafe-packages-json-L933

File: packages/core/database/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (smorimoto) [reply packages/core/database/package.json:40]: Anytime, Indeed.

File: packages/core/helper-plugin/package.json (hunk 1 above)
Comment (smorimoto) [reply packages/core/helper-plugin/package.json]: Done.

File: packages/generators/generators/package.json (hunk 2 above)
Comment (smorimoto) [reply packages/generators/generators/package.json]: Done.

File: packages/plugins/i18n/package.json (hunk 3 above)
Comment (smorimoto) [reply packages/plugins/i18n/package.json]: Done.

**Comment by smorimoto** [comment]
@joshuaellis I also reverted all bin and other normalise changes. Once it's ready to merge, I will clean up all the commits.

**Review by joshuaellis — COMMENTED** [review]
No more changes from my end, think you could squash those commits?

**Comment by smorimoto** [comment]
@joshuaellis Fixed all conflicts then rebased to the latest main branch 🙂

**Comment by joshuaellis** [comment]
Looks like the linter is broken – have you tried running it locally?

**Comment by smorimoto** [comment]
@joshuaellis Weird. I just fixed the lockfile.

**Comment by joshuaellis** [comment]
Hey @Convly can we please get an experimental release off this PR?

----

Great work @smorimoto, lets get an experimental release to ensure there aren't any adverse affects in what we publish – locally I couldn't see any issues and it _did_ build for me, so 🤞🏼

**Comment by smorimoto** [comment]
Fantastic! Thanks again for your very quick and very detailed review!

**Comment by derrickmehaffy** [comment]
> Hey @Convly can we please get an experimental release off this PR?
> 
> ----
> 
> Great work @smorimoto, lets get an experimental release to ensure there aren't any adverse affects in what we publish – locally I couldn't see any issues and it _did_ build for me, so 🤞🏼 

We can't do nightlies on branches that don't exist within our repo and I don't see it so I assume it's coming from a fork. We would need to merge it into a branch on our side.

(At least we can't using the GH action for nightly, it could of course be done manually)

**Comment by gu-stav** [comment]
I guess we could point this PR to strapi-branch then, merge it and re-open the PR? An experimental release would be really good for this one. 👍🏼

**Comment by joshuaellis** [comment]
> I guess we could point this PR to strapi-branch then, merge it and re-open the PR? An experimental release would be really good for this one. 👍🏼

Let's see what @Convly wants us to do.

**Comment by derrickmehaffy** [comment]
> I guess we could point this PR to strapi-branch then, merge it and re-open the PR? An experimental release would be really good for this one. 👍🏼

Yeah we can create a temp internal branch and change where it's pointed and merge it there. Agree though lets see what JS wants to do. 

@gu-stav / @joshuaellis do you want me to record a short loom video showing how you can generate a nightly from a branch? (It's quite easy)

**Comment by joshuaellis** [comment]
> @gu-stav / @joshuaellis do you want me to record a short loom video showing how you can generate a nightly from a branch? (It's quite easy)

I'm assuming therefore we don't need publishing rights to NPM? In which case, that would be super useful yes please.

**Comment by derrickmehaffy** [comment]
> > @gu-stav / @joshuaellis do you want me to record a short loom video showing how you can generate a nightly from a branch? (It's quite easy)
> 
> I'm assuming therefore we don't need publishing rights to NPM? In which case, that would be super useful yes please.

Nope you don't, it's all handled inside of GH actions. I'll record a loom and share it via Slack.
